dlt/common/configuration/providers/toml.py

```diff
@@ -113,124 +132,49 @@ def name(self) -> str:
-    The basic working principle is obtain config and secrets values from Vault keys and reconstitute a `secrets.toml` like document that is then used
-    as a cache.
-
-    The implemented must provide `_look_vault` method that returns a value from external vault from external key.
-
-    To reduce number of calls to external vaults the provider is searching for a known configuration fragments which should be toml documents and merging
-    them with the
-    - only keys with secret type hint (CredentialsConfiguration, TSecretValue) will be looked up by default.
-    - provider gathers `toml` document fragments that contain source and destination credentials in path specified below
-    - single values will not be retrieved, only toml fragments by default
-
-    """
-
-    def __init__(self, only_secrets: bool, only_toml_fragments: bool) -> None:
-        """Initializes the toml backed Vault provider by loading a toml fragment from `dlt_secrets_toml` key and using it as initial configuration.
+class CustomLoaderDocProvider(BaseDocProvider):
+    def __init__(
+        self, name: str, loader: Callable[[], Dict[str, Any]], supports_secrets: bool = True
+    ) -> None:
+        """Provider that calls `loader` function to get a Python dict with config/secret values to be queried.
+        The `loader` function typically loads a string (ie. from file), parses it (ie. as toml or yaml), does additional
+        processing and returns a Python dict to be queried.
 
-        _extended_summary_
+        Instance of CustomLoaderDocProvider must be registered for the returned dict to be used to resolve config values.
+        >>> import dlt
+        >>> dlt.config.register_provider(provider)
 
         Args:
-            only_secrets (bool): Only looks for secret values (CredentialsConfiguration, TSecretValue) by returning None (not found)
-            only_toml_fragments (bool): Only load the known toml fragments and ignore any other lookups by returning None (not found)
-        """
-        self.only_secrets = only_secrets
-        self.only_toml_fragments = only_toml_fragments
-        self._vault_lookups: Dict[str, pendulum.DateTime] = {}
+            name(str): name of the provider that will be visible ie. in exceptions
+            loader(Callable[[], Dict[str, Any]]): user-supplied function that will load the document with config/secret values
+            supports_secrets(bool): allows to store secret values in this provider
 
-        super().__init__(tomlkit.document())
-        self._update_from_vault(SECRETS_TOML_KEY, None, AnyType, None, ())
-
-    def get_value(
-        self, key: str, hint: type, pipeline_name: str, *sections: str
-    ) -> Tuple[Optional[Any], str]:
-        full_key = self.get_key_name(key, pipeline_name, *sections)
+        """
+        self._name = name
+        self._supports_secrets = supports_secrets
+        super().__init__(loader())
```

Comment:
What is the use of this CustomLoaderProvider if it is not lazily evaluated on first access of the dict?

Reply:
heh you are right. in this implementation it does not make sense. BUT maybe we want to change it to lazy evaluation ie. on first use. may be necessary ie. on Airflow. so let's keep it like that!